[t1230] iframe dialogs removed

secretrobotron opened the pull request:
1. Wrote new dialog.js to replace IFrameDialog and WindowDialog
2. Removed IFrameDialog and WindowDialog
3. Morphed IFrameDialog slightly specifically for editor module
4. Moved dialogs into src/
5. Ported dialog iframe code to new system
6. Fixed up html/css
7. Fixed try/catch in cornfield which caused debug hilarity and confusion

secretrobotron committed:
[t1230] iframe dialogs removed

1. Wrote new dialog.js to replace IFrameDialog and WindowDialog
2. Removed IFrameDialog and WindowDialog
3. Morphed IFrameDialog slightly specifically for editor module
4. Moved dialogs into src/
5. Ported dialog iframe code to new system
6. Fixed up html/css
7. Fixed try/catch in cornfield which caused debug hilarity and confusion

secretrobotron committed:
[t1230] lint/make fixes

dseif commented:
/botio check

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/ae59d0ee3ebbd9d/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/ae59d0ee3ebbd9d/output.txt

Total script time: 1.33 mins

dseif commented:
Don't know why the robot is whining, tested locally and it works fine

dseif commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,186 @@
+        if ( _enterKeyActivity && __keyboardAvoidElements.indexOf( e.target.nodeName ) === -1 && ( e.which === 13 || e.keyCode === 13 ) ) {
+          _activities[ _enterKeyActivity ]( e );
+        }
+        else if ( _escapeKeyActivity && __keyboardAvoidElements.indexOf( e.target.nodeName ) === -1 && ( e.which === 27 || e.keyCode === 27 ) ) {
```
can we split this line and line 27 up a bit

dseif commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,186 @@
+      }
+
+      var _internal = {
+        rootElement: _rootElement,
```
does this ever get used? Looks like we can just use _rootElement

dseif replied:
never mind I see it being used later in the actual dialogs, my bad.

dseif commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,186 @@
+        },
+
+        disableElements: function(){
+          for( var i=0; i<arguments.length; ++i ){
```
can store a reference to arguments.length for some extra goodness, on line 81 as well

secretrobotron replied:
Two things to consider here:

1) I didn't rewrite this code, I only ported it, mostly through copy & paste.
2) While I do like optimizing, these are probably instances where it will just become more confusing to read, rather than add a whole lot of value, no?

dseif replied:
If you didn't write the code then it's not a huge deal.

humphd replied:
Caching arguments.length is advised, yes.  Whether you wrote this code or not, you've touched it now, so we have to deal with it here.

dseif commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,186 @@
+          _modal = new Modal( _rootElement );
+          setTimeout( function() {
+            _external.focus();
+          }, 0 );
```
Why do we need to do the setTimeout 0, is there a way around this?

secretrobotron replied:
Like I said in the comment above, I've just copied and pasted most of this code. However, I can see why this might be here: to let the user-agent do what it needs to do before applying the changes from `focus()`. I think we added this to prevent some problems with the user-agent grabbing focus improperly otherwise. Regardless, I don't see problems with these kind of `setTimout` blocks because they're usually UI-oriented.

dseif replied:
Same as above, if you didn't write the code then this is fine.

humphd replied:
This is exactly why I wanted async events in the EventMangerWrapper stuff I did.  By having an async event you solve this issue.  Firefox has gotten into a lot of problem using this hack, see http://blog.mozilla.org/tglek/2012/04/10/argh-at-our-unresponsive-tab-strip-settimeoutfoo-0-can-be-very-harmful/  I think we need to move to async event for this, and drop setTimeout.  That will need #1130 to land, so if you want to file a bug blocked on that, I can fix it later.

dseif commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,186 @@
+      // make this happen a tiny bit later so variables get initialized before opening
+      setTimeout( function(){
+        _external.open();
+      }, 10 );
```
anyway we could do this in a callback instead?

secretrobotron replied:
How do you mean?

dseif replied:
Is there a way to call `_external.open(); after the variables get initialized? ( reading comment ). If this is another case of you not writing this code then ignore this comment.

secretrobotron replied:
No, I think you're right in pointing this out, but I'm not sure it's bad/wrong. My intent was for you to be able to call `spawn()` and just have the dialog appear without waiting for it to initialize. Think that's too confusing?

dseif replied:
I like the idea of calling `spawn()` and it opening the dialog for us, I think that's awesome. I think the setTimeout bothers me because 10 milliseconds just seems like an arbitrary number and gives us no guarantee that everything will be ready in time.

secretrobotron replied:
What about this? https://developer.mozilla.org/en/DOM/window.setTimeout#Minimum_delay_and_timeout_nesting

I just picked `10` because it's ... yeah. Probably should be `0`.

dseif replied:
Alright cool if this is just being done so it doesn't block the main thread then +1. After changing it to 0 this can go to SR

dseif commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,186 @@
+        return __openDialogs[ name ];
+      }
+      else {
+        throw "Dialog '" + name + "' does no exist.";
```
should this be throw new error?

secretrobotron replied:
I thought so too, at first, but it doesn't look like it's necessary: https://developer.mozilla.org/en/JavaScript/Reference/Statements/throw

dseif replied:
Cool

dseif commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,186 @@
+        throw "Dialog '" + name + "' does no exist.";
+      }
+    }
+
```
get rid of this whitespace

dseif commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,186 @@
+    }
+
+  };
+
```
same here

dseif commented on src/dialog/dialogs/delete-track.html:
```diff
@@ -0,0 +1,16 @@
+    <button class="no">No</button>
+  </div>
+  <div class="close-button"></div>
+</div>
```
This is soooo much better than iframes, is the plan to move the editors to something like this as well?

secretrobotron replied:
Yep :D.

dseif commented on src/dialog/dialogs/export.js:
```diff
@@ -0,0 +1,44 @@
+        htmlButton = dialog.rootElement.querySelector( ".html-button" ),
+        jsonExport = dialog.rootElement.querySelector( ".json-export" ),
+        htmlExport = dialog.rootElement.querySelector( ".html-export" ),
+        title = dialog.rootElement.querySelector( ".title" );
```
can you create a reference to dialog.rootElement

dseif commented on src/dialog/dialogs/delete-track.js:
```diff
@@ -0,0 +1,23 @@
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+define([ "text!dialog/dialogs/delete-track.html", "dialog/dialog" ],
```
can you explain the text!dialog part of this, I don't get what it's doing

dseif replied:
does that mean it get's returned as plain text as opposed to an object?

secretrobotron replied:
Yeah, it's how the layout system works. `text!` will deliver a piece of text and not try to run it as code. Great for including HTML in your projects :D.

dseif replied:
Yea that's pretty awesome

dseif commented on src/dialog/dialogs/load.js:
```diff
@@ -0,0 +1,31 @@
+      });
+    });
+
+    for( var i=0; i<projects.length; ++i ){
```
store reference to projects.length

secretrobotron replied:
Again, I'd like to make the case that readability is good here. This doesn't need to be a hot loop, right?

dseif replied:
I dunno `i = 0, l = projects.length; i < l;` is still pretty readable. Hot loop? If this is another case of you not writing this code then ignore it.

humphd replied:
Have to agree with Dave, cache, cache, cache.

dseif commented on src/dialog/dialogs/save-as.html:
```diff
@@ -0,0 +1,16 @@
+      obtain one at http://www.mozillapopcorn.org/butter-license.txt -->
+
+<div class="butter-dialog">
+  <h1>Please give your project a<span class="better"> better</span> name:</h1>
```
haha what is the span there for?

secretrobotron replied:
Try saving with a blank filename :). Again, copy & pasted.

dseif commented on src/dialog/dialogs/save-as.js:
```diff
@@ -0,0 +1,27 @@
+        dialog.send( "submit", nameInput.value );
+      }
+      else{
+        dialog.rootElement.querySelector( ".better" ).style.display = "inline";
```
ah I see what that span was being used for now

secretrobotron replied:
Perfect :D.

dseif commented on src/dialog/dialogs/track-data.js:
```diff
@@ -0,0 +1,37 @@
+
+  Dialog.register( "track-data", LAYOUT_SRC, function ( dialog, track ) {
+    var trackName = dialog.rootElement.querySelector( ".track-name" ),
+        trackData = dialog.rootElement.querySelector( ".track-data" );
```
store reference to rootElement please

dseif commented on src/dialog/modal.js:
```diff
@@ -6,34 +6,30 @@ define( [], function(){
     // ow, opacity transition fails to render
     setTimeout( function(){
-      _element.className += " fade-in";
+      _element.classList.add( "fade-in" );
```
we should probably use our shim for this so once we add IE9 support we don't run into issues

secretrobotron replied:
It should already be included, actually. To be absolutely certain though probe Chris about the status of this, https://webmademovies.lighthouseapp.com/projects/65733/tickets/1231-insert-shims-on-startup.

dseif replied:
Yea I stole the ticket, will review after this

dseif commented on src/dialogs.js:
```diff
@@ -0,0 +1,25 @@
+  ];
+
+  var include = [];
+  for ( var i = 0; i < dialogs.length; ++i ) {
```
store reference to dialogs.length

secretrobotron replied:
Same as above. This gets run once, and there aren't thousands of dialogs. Hot loop necessary?

humphd replied:
Even better:

var i = dialogs.length;
while( i-- ){
  ... dialogs[ i ]...
}

secretrobotron replied:
Yeah, that's a lot nicer. I'm curious about why is this an issue to fix up here though? Should we always ensure that lengths are cached in loops during review? I like being speedy, but catching all of these seems more pedantic than necessary imo.

Happy to change it regardless.

dseif commented on src/ui/header.js:
```diff
@@ -129,14 +110,9 @@ define( [ "dialog/iframe-dialog" ], function( IFrameDialog ){
-        type: "iframe",
-        modal: true,
-        url: butter.ui.dialogDir + "error-message.html",
+      var dialog =Dialog.spawn( "error-message", {
```
add space after =

dseif commented:
When you open the load dialog, the "x" button in the corner is just a blue box

dseif commented:
Going to need tests for this, don't know if you want to write them now or do what Humph mentioned in an earlier email and stage this then set a tag to "needs-tests" or whatever.

secretrobotron commented:
Aye. I'll add tests.

secretrobotron commented:
Did you get that blue box in Chrome? FF? Both? It happened for me too, but my gfx card was already flipping out. Couldn't tell if it was just Chrome or not.

secretrobotron committed:
[t1230] review fixes (mostly spacing/caching)

secretrobotron committed:
[t1230] more spacing issues for review

secretrobotron committed:
[t1230] tests and a couple fixes to help them run

secretrobotron committed:
[t1230] lint fix

dseif commented:
It was in Chrome that I got the blue box, will try again and see if I can reproduce it.

dseif commented:
/botio check

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/973f183196eacff/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/973f183196eacff/output.txt

Total script time: 0.33 mins

dseif commented:
/botio preview

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/a3a6cb4ed58814e/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/a3a6cb4ed58814e/output.txt

Total script time: 0.62 mins

Server is running on http://pr363.butyr.org/templates/test.html

dseif commented:
Yea the blue box issue on the load dialog looks to be fixed

secretrobotron committed:
[t1230] changed setTimeout time to 0

dseif commented:
/botio preview

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/dfb520adf3d4353/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/dfb520adf3d4353/output.txt

Total script time: 0.93 mins

Server is running on http://pr363.butyr.org/templates/test.html

humphd commented on css/dialog.css:
```diff
@@ -61,11 +34,11 @@ html, body {
   top: 5px;
   right: 5px;
-  background-image: url( "assets/close-button.png" );
+  background-image: url( "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAC4AAAArCAYAAAAHdLqEAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAylJREFUeNrUWTFvEzEUdiIWkFoVtUuhlZhgYLkwIMGSIpAYEJAOULYmUvgBRazQCz8gaRYYiJSObVkCqAMSqO0AlZAgtzCUKVKLuoAoBcEY3guv1dlNuGef7675JOsS587+/PL58zs71W63RT/iiI1GvryvnYJLDkoWCn52utzmUVmD0jh5vrgTps9UmIgD4Qm4zEKZMHi8AaUKA1iNjTgQxohWDAmrQOIFGEBL56G0AWkXLk1LpAW104R285F ...
```
Why not let the browser cache this file?  With icons we load as separate files, why not here?

humphd commented on css/media-view.css:
```diff
@@ -0,0 +1,55 @@
+}
+.butter-media-properties .url input,
+.butter-media-properties .url button {
+  transition: opacity 0.2s;
```
Needs prefix versions.

humphd commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,197 @@
+        enableCloseButton: function(){
+          var closeButton = _rootElement.querySelector( ".close-button" );
+          if( closeButton ){
+            closeButton.addEventListener( "click", function( e ){
```
Remove this on fire?

closeButton.addEventListener( "click", function clickHandler( e ){
this.removeEventListener( "click", clickHandler, false);
...

humphd commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,197 @@
+
+        assignButton: function( selector, activityName ){
+          var element = _rootElement.querySelector( selector );
+          element.addEventListener( "click", _activities[ activityName ], false );
```
Same deal here, won't we leak this?

secretrobotron replied:
I'm really not entirely sure. While I'll remove the listener for the sake of good practice, will this just get GC'd since references removed completely?

humphd replied:
No, let's leave it.  Based on irc discussion with you/rick, I think it's fine.  My hesitation comes from historical issues.

humphd commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,197 @@
+      dialogCtor( _internal, spawnOptions.data );
+
+      // make this happen a tiny bit later so variables get initialized before opening
+      setTimeout( function(){
```
Same deal here, we want an async event instead.

humphd commented on src/dialog/dialogs/error-message.js:
```diff
@@ -0,0 +1,15 @@
+
+  Dialog.register( "error-message", LAYOUT_SRC, function( dialog, data ) {
+    var message = dialog.rootElement.querySelector( ".message" );
+    message.innerHTML = data;
```
Kind of sucks that you still have to work with innerHTML here.  Should we provide some nicer way to not have to work directly with a dom node?

secretrobotron replied:
I agree. However, I think that's another enhancement on top of this ticket. File a new one?

humphd commented on src/dialog/dialogs/export.js:
```diff
@@ -0,0 +1,46 @@
+        htmlExport = rootElement.querySelector( ".html-export" ),
+        title = rootElement.querySelector( ".title" );
+
+    title.innerHTML = "HTML Export";
```
Same thing here...I'm wondering if we want to be working against a "dialog" object vs. a dialog-html-fragment?

humphd commented on src/dialog/modal.js:
```diff
@@ -4,36 +4,31 @@
+    _element.appendChild( childElement );
+
     this.destroy = function(){
       __container.removeChild( _element );
```
also _element = null;

humphd commented on src/main.js:
```diff
@@ -76,6 +80,8 @@
 
       EventManagerWrapper( _this );
 
+      this.dialog = Dialog;
```
Will you later just get this via require instead of holding an instance?

secretrobotron replied:
It was my intent to leave it open for people to use `butter.dialog.register()` during execution instead of having to have your whole stack during compile time.

humphd replied:
OK, that's awesome.  Maybe leave a comment to that effect above this line.

secretrobotron committed:
[t1230] review changes

secretrobotron committed:
[t1230] rebase, review changes, cleanup

1. Rebased off of master
2. Used while() instead of for(i) for quick loops
3. Dialogs must call open() after spawn() to actually open
4. css fixes, cleanup, review changes

humphd commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,197 @@
+          _escapeKeyActivity,
+          _modal;
+
+      // make sure have a handle to the butter-dialog div
```
Typo: "make sure ^we have a handle..."

humphd commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,197 @@
+        }
+      }
+
+      var _internal = {
```
Please add an explanatory comment to _internal and _external, and how they are used.  In general, this code needs a bit more documentation.

humphd commented on src/dialog/dialog.js:
```diff
@@ -0,0 +1,197 @@
+        }
+      };
+
+      var _external = {
```
Same deal here.  Explain a bit.

secretrobotron committed:
[t1230] documentation

jbuck commented:
Staged on master: https://github.com/mozilla/butter/compare/510b88cd8039...bd9990522eb2
